content/docs/reference-react-component.md

```diff
@@ -311,21 +313,110 @@ In the above examples, it is important to read the `scrollHeight` property in `g
+    super(props);
+    this.state = { hasError: false };
+  }
+
+  static getDerivedStateFromError(error) {
+    // Update state so the next render will show the fallback UI.
+    return { hasError: true };
+  }
+
+  render() {
+    if (this.state.hasError) {
+      // You can render any custom fallback UI
+      return <h1>Something went wrong.</h1>;
+    }
+
+    return this.props.children; 
+  }
+}
+```
+
+> Note
+>
+> `getDerivedStateFromError()` is called during the "render" phase, so side-effects are not permitted.
+For those use cases, use `componentDidCatch()` instead.
+
+* * *
+
 ### `componentDidCatch()`
 
 ```javascript
 componentDidCatch(error, info)
 ```
 
-[Error boundaries](/docs/error-boundaries.html) are React components that catch JavaScript errors anywhere in their child component tree, log those errors, and display a fallback UI instead of the component tree that crashed. Error boundaries catch errors during rendering, in lifecycle methods, and in constructors of the whole tree below them.
+This lifecycle is invoked after an error has been thrown by a descendant component.
+It receives two parameters:
 
-A class component becomes an error boundary if it defines this lifecycle method. Calling `setState()` in it lets you capture an unhandled JavaScript error in the below tree and display a fallback UI. Only use error boundaries for recovering from unexpected exceptions; don't try to use them for control flow.
+1. `error` - The error that was thrown.
+2. `info` - An object with a `componentStack` key containing [information about which component threw the error](/docs/error-boundaries.html#component-stack-traces).
 
-For more details, see [*Error Handling in React 16*](/blog/2017/07/26/error-handling-in-react-16.html).
+
+`componentDidCatch()` is called during the "commit" phase, so side-effects are permitted.
+It should be used for things like logging errors:
+
+```js{12-19}
+class ErrorBoundary extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { hasError: false };
+  }
+
+  static getDerivedStateFromError(error) {
+    // Update state so the next render will show the fallback UI.
+    return { hasError: true };
+  }
+
+  componentDidCatch(error, info) {
+    // Example "componentStack":
+    //   in ComponentThatThrows (created by App)
+    //   in ErrorBoundary (created by App)
+    //   in div (created by App)
+    //   in App
+    logComponentStackToMyService(info.componentStack);
+  }
+
+  render() {
+    if (this.state.hasError) {
+      // You can render any custom fallback UI
+      return <h1>Something went wrong.</h1>;
+    }
+
+    return this.props.children; 
+  }
+}
+```
 
 > Note
 > 
-> Error boundaries only catch errors in the components **below** them in the tree. An error boundary can’t catch an error within itself.
+> In the event of an error, you can render a fallback UI with `componentDidCatch()` by calling `setState`, but this will be deprecated in a future release.
```

Comment:
what are advantages of separating those 2 things? i understand the part about side effects being allowed in one and not in the other one, but `componentDidCatch` is working for both cases at the moment, so what's the reason behind separating them? non of them should really be called often (they are last resort kind of thing) so it raises a question if they are both needed as they will for sure be confusing for newcomers

Reply:
My understanding of this is that there are a couple of motivating factors:

**It works with server-side rendering**. `componentDidCatch` is a _commit phase_ lifecycle, but there is no commit phase on the server. `getDerivedStateFromError` is a _render phase_ lifecycle, and so it can be used to enable error handling on the server.

**Render phase recovery is safer**. The story for error recovery via `componentDidCatch` is a little janky, since it relies on an intermediate commit of "null" for everything below the component that errored. This might result in subsequent errors inside of any components higher up in the tree that implement `componentDidMount` or `componentDidUpdate` and _just assume_ that their refs will be non-null (because they always are in the non-error case).

**It's more optimal for async rendering**. Because state-updates from commit phase lifecycles are _always synchronous_, and because `componentDidCatch` is called during the commit phase– using `componentDidCatch` for error recovery is not optimal because it forces the fallback UI to always render synchronously.

Reply:
I ofc dont judge the chosen APIs too harshly as Im not competent enough to do that, not knowing all intricacies of the implementation. Just commenting from the user's perspective. And thanks for the explanation!

Just want to comment a little bit on the given arguments

- **It works with server-side rendering** - this is appealing
- **Render phase recovery is safer.** - seems like implementation detail that could be handled differently, but i understand it would probably make the implementation more clunky as it would have to be special cased commit hook
- **It's more optimal for async rendering.** - not the strongest argument (as you have mentioned - those should be edge cases and thus being optimal for rendering shouldnt matter as much)